File: Chicago/chicago_prep_data.py
import pandas as pd
import numpy as np
from datetime import datetime

# It's much faster to read the file if we fix the date/time format.
# Otherwise it has to guess each row's format.
to_datetime = lambda d: datetime.strptime(d, '%m/%d/%Y %H:%M:%S %p')

# Since the file is large, load a limited number of variables.
c = ["Date", "Primary Type", "Community Area", "District", "Ward", "Latitude", "Longitude"]
df = pd.read_csv("chicago.csv.gz", usecols=c, converters={"Date": to_datetime})
df = df.rename(columns={"Community Area": "CommunityArea", "Primary Type": "PrimaryType"})

# Limit to the 10 most common types of crime
ptype = df.loc[:, "PrimaryType"]
pt = ptype.value_counts()
pt10 = pt[0:10].index
dx = df.loc[ptype.isin(pt10), :]

# Count the number of times each crime type occurs in each community area on each day.
first = lambda x: x.iloc[0]
dx["DateOnly"] = dx.Date.dt.date
dy = dx.groupby(["PrimaryType", "DateOnly", "CommunityArea"]).agg({"DateOnly": [np.size, lambda x: x.iloc[0]],
                 "District": first, "Ward": first, "Latitude": first, "Longitude": first})

# Expand the index so that every day has a record
a = dx["PrimaryType"].unique()
b = np.arange(dx.DateOnly.min(), dx.DateOnly.max())
c = dx["CommunityArea"].unique()
ix = pd.MultiIndex.from_product([a, b, c])
dy = dy.reindex(ix)

dy.columns = [' '.join(col).strip() for col in dy.columns.values]
dy = dy.rename(columns={'Latitude <lambda>': "Latitude",
                        'Ward <lambda>': "Ward",
                        'DateOnly size': "Num",
                        'DateOnly <lambda>': "DateOnly",
                        'Longitude <lambda>': "Longitude",
                        'District <lambda>': "District"})

dy.Num = dy.Num.fillna(0)

# Move the hierarchical row index into columns, rename to get more meaningul names.
cdat = dy.reset_index()
cdat = cdat.rename(columns={"level_0": "PrimaryType", "level_1": "Date", "level_2": "CommunityArea"})

# Split date into variables for year, day within year, and day of week.
cdat["DayOfYear"] = cdat.Date.dt.dayofyear
cdat["Year"] = cdat.Date.dt.year
cdat["DayOfWeekNum"] = cdat.Date.dt.dayofweek
cdat["DayOfWeek"] = cdat.Date.dt.dayofweek.replace(
    {0: "Mo", 1: "Tu", 2: "We", 3: "Th", 4: "Fr", 5: "Sa", 6: "Su"})

cdat.to_csv("cdat.csv.gz", compression="gzip")

File: Chicago/chicago_prep_data.R
library(data.table)
library(lubridate)
library(R.utils)

# Since the file is large, load a limited number of variables.
c = c("Date", "Primary Type", "Community Area")
df = fread("zcat chicago.csv.gz", select=c, showProgress=TRUE)
setnames(df, old="Community Area", new="CommunityArea")
setnames(df, old="Primary Type", new="PrimaryType")

# Convert the date (we don't need the time)
setnames(df, old="Date", new="DateString")
df[, Datect := as.POSIXct(DateString, format="%m/%d/%Y %H:%M:%S %p", tz="America/Chicago")]
df[, Date := as.Date(Datect)]

# Get rid of columns that we no longer need
df[, Datect := NULL]
df[, DateString := NULL]

# Limit to the 10 most common types of crime
ft = table(df$PrimaryType)
ft = sort(ft, decreasing=TRUE)[1:10]
ft = names(ft)
df = df[PrimaryType %in% ft,]

# Count the number of times each crime type occurs in each community area on each day
setkey(df, "Date", "PrimaryType", "CommunityArea")
df[, Num := .N, by=c("Date", "PrimaryType", "CommunityArea")]
dc = df[, head(.SD, 1), by=c("Date", "PrimaryType", "CommunityArea")]

# Create a complete enumeration of location x date x crime type
da = seq(from=min(dc$Date), to=max(dc$Date), "days")
ix = CJ(Date=da, PrimaryType=unique(dc$PrimaryType), CommunityArea=unique(dc$CommunityArea))

# Include zeros where no crimes were reported
setkey(dc, "Date", "PrimaryType", "CommunityArea")
setkey(ix, "Date", "PrimaryType", "CommunityArea")
dc = dc[ix]
j = which(names(dc) == "Num")
set(dc, which(is.na(dc$Num)), j, 0)

# Construct values for year, day of year, and day of week
dc[, Year := lubridate::year(Date)]
dc[, DayOfYear := lubridate::yday(Date)]
dc[, DayOfWeek := lubridate::wday(Date)]

dc = na.omit(dc)

fwrite(dc, "cdat_R.csv")
gzip("cdat_R.csv", "cdat_R.csv.gz", overwrite=TRUE)
